src/scene/layers.js

```diff
@@ -0,0 +1,786 @@
+pc.extend(pc, function () {
+    var keyA, keyB, sortPos, sortDir;
+
+    function sortManual(drawCallA, drawCallB) {
+        return drawCallA.drawOrder - drawCallB.drawOrder;
+    }
+
+    function sortMaterialMesh(drawCallA, drawCallB) {
+        keyA = drawCallA._key[pc.SORTKEY_FORWARD];
+        keyB = drawCallB._key[pc.SORTKEY_FORWARD];
+        if (keyA === keyB && drawCallA.mesh && drawCallB.mesh) {
+            return drawCallB.mesh.id - drawCallA.mesh.id;
+        }
+        return keyB - keyA;
+    }
+
+    function sortBackToFront(drawCallA, drawCallB) {
+        return drawCallB.zdist - drawCallA.zdist;
+    }
+
+    function sortFrontToBack(drawCallA, drawCallB) {
+        return drawCallA.zdist - drawCallB.zdist;
+    }
+
+    var sortCallbacks = [null, sortManual, sortMaterialMesh, sortBackToFront, sortFrontToBack];
+
+    function sortCameras(camA, camB) {
+        return camA.priority - camB.priority;
+    }
+
+    var _guidA, _guidB;
+    function sortLights(lightA, lightB) {
+        _guidA = lightA._node._guid;
+        if (!_guidA) _guidA = 0; // does it make sense?
+        
+        _guidB = lightB._node._guid;
+        if (!_guidB) _guidB = 0;
+
+        return _guidA.localeCompare(_guidB);
+    }
+
+    // Layers
+    var layerCounter = 0;
+    var layerList = [];
+
+    var VisibleInstanceList = function () {
+        this.list = [];
+        this.length = 0;
+        this.done = false;
+    };
+
+    var InstanceList = function () {
+        this.opaqueMeshInstances = [];
+        this.transparentMeshInstances = [];
+        this.shadowCasters = [];
+
+        // arrays of VisibleInstanceList for each camera
+        this.visibleOpaque = [];
+        this.visibleTransparent = [];
+    };
+
+    var Layer = function (options) {
```

Comment:
Need to document all these classes and methods (and mark private?)

Reply:
Done.